sled-agent/src/illumos/dladm.rs

```diff
@@ -164,15 +176,24 @@ impl Dladm {
         Ok(())
     }
 
-    /// Returns all VNICs that may be managed by the Sled Agent.
-    pub fn get_vnics() -> Result<Vec<String>, GetVnicError> {
+    /// Returns VNICs that may be managed by the Sled Agent, optionally
+    /// restricted to a particular kind.
+    pub fn get_vnics(
+        kind: Option<VnicKind>,
+    ) -> Result<Vec<String>, GetVnicError> {
         let mut command = std::process::Command::new(PFEXEC);
         let cmd = command.args(&[DLADM, "show-vnic", "-p", "-o", "LINK"]);
         let output = execute(cmd).map_err(|err| GetVnicError { err })?;
 
         let vnics = String::from_utf8_lossy(&output.stdout)
             .lines()
-            .filter(|vnic| vnic.starts_with(VNIC_PREFIX))
+            .filter(|name| {
+                if let Some(kind) = kind {
+                    VnicKind::from_name(name) == kind
+                } else {
+                    false
+                }
```

Comment:
So if a caller passes `kind == None`, we'll always return `false`.

Why bother passing an option? Why not just directly supply `VnicKind` as an argument?

Reply:
Ahh, I meant that to return `true`. The goal was to have `None` be equivalent to "get me all VNICs on the system, of any kind".

Reply:
Do we ever want that though? Like, the old version of this function only returned VNICs that the sled agent knew about. The new version may return vnics that aren't being managed by Oxide too, right?

Reply:
That's true. I think one improvement would be to remove the `VnicKind::Other` variant, and never return VNIC names that don't correspond to one of the known naming conventions. So, I guess we really only want `VnicKind::{OxideControl,Guest}`. `XdeUnderlay` will be managed by the `{create,destroy}_virtual_hardware.sh` scripts, not the sled agent. And `VnicKind::Other` are off limits entirely.

That seems better, and maybe also accept only a single `VnicKind` argument, not an `Option`, as you're suggesting. Thoughts?

Reply:
Sounds good to me - that accomplishes the goal I was aiming for, which is "make it hard to accidentally mess with state not owned by the Oxide stack".

Reply:
Sweet, will do both. Thanks!

Reply:
Resolved in [863b5ca](https://github.com/oxidecomputer/omicron/pull/1052/commits/863b5cabe1a88603ee1ef515eba4b8ba464d831e)